### PR TITLE
Add support for TF_VAR_ environment variables and .env files

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
This is the PR with @HadwaAbdelhalem. 

According to the conversation of this issue, I implement the feature. 

https://github.com/charleszipp/azure-pipelines-tasks-terraform/issues/97

This feature enable to read secureFile as `.env` file and parse it and makes it available as  environment variables. I tested on Azure DevOps and works fine. The secure file name is `*.env` it is  referred  as `.env` file and remove `-var-file` option.

The test code will be created by the following PR. 
(When I wrote the test, the test add `-var-file` option even if it is `.env` file. It works on Azure DevOps.  However, I can't identify where it happens. If you know where it is, please tell me. :)

### CODE REVIEW

**Comment by TsuyoshiUshio** [comment]
@charleszipp FYI, all these comment  is done already before creating PR. Ignore it.

**Comment by TsuyoshiUshio** [comment]
Hi @charleszipp Could you share with me the root cause of pipeline failure, please? 
or could you add me to the AzDO? I'm a MSFT guy. Tsuyoshi Ushio.

**Comment by rguthriemsft** [comment]
@TsuyoshiUshio try your access again.  I added you to project team.

**Comment by TsuyoshiUshio** [comment]
Hi @charleszipp , @rguthriemsft , and @HadwaAbdelhalem 

I checked the CI and run `npm run test` locally. It works.  I found this log on the CI. Any thoughts about it? I keep on investigating the root cause. 

```
2020-02-07T00:07:05.0445306Z [command]C:\windows\system32\cmd.exe /D /S /C ""C:\Program Files\nodejs\npm.cmd" run test"
2020-02-07T00:07:07.0744780Z internal/modules/cjs/loader.js:796
2020-02-07T00:07:07.0745997Z     throw err;
2020-02-07T00:07:07.0746430Z     ^
2020-02-07T00:07:07.0746770Z 
2020-02-07T00:07:07.0746985Z Error: Cannot find module 'D:\a\1\s\TerraformCLI\node'
2020-02-07T00:07:07.0747205Z     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:793:17)
2020-02-07T00:07:07.0747814Z     at Function.Module._load (internal/modules/cjs/loader.js:686:27)
2020-02-07T00:07:07.0748057Z     at Function.Module.runMain (internal/modules/cjs/loader.js:1043:10)
2020-02-07T00:07:07.0748264Z     at internal/main/run_main_module.js:17:11 {
2020-02-07T00:07:07.0748475Z   code: 'MODULE_NOT_FOUND',
2020-02-07T00:07:07.0748667Z   requireStack: []
2020-02-07T00:07:07.0748872Z }
```

**Comment by charleszipp** [comment]
@TsuyoshiUshio it looks like there was a [defect](https://github.com/istanbuljs/nyc/issues/1205) in the version of nyc that was installed in the project. I was able to reproduce what the pipeline was reporting locally immediately after cloning the repo and running `npm run test`. 

After i upgraded to latest version of nyc, I was no longer able to repro. Waiting on pipeline to rerun to confirm that resolved the issue

File: README.md [hunk 1]
```diff
@@ -12,6 +12,10 @@ See readme for each of the tasks for development setup for each.
 
 ## Release Notes
 
+### 0.4.15
```
Comment (charleszipp) [inline README.md]: The next version will be 0.4.16

File: TerraformCLI/src/terraform-validate.ts [hunk 2]
```diff
@@ -6,12 +6,12 @@ import { TerraformRunner } from "./terraform-runner";
...
 export class TerraformValidate extends TerraformCommand{
     readonly secureVarsFile: string | undefined;
-
     constructor(
         name: string, 
         workingDirectory: string,
         options?: string, 
-        secureVarsFile?: string) {
+        secureVarsFile?: string,
+        isEnvFile?: boolean) {
```
Comment (charleszipp) [inline TerraformCLI/src/terraform-validate.ts]: This var is never read. Can we remove it?

File: TerraformCLI/src/terraform-runner.ts [hunk 3]
```diff
@@ -85,9 +87,24 @@ export class TerraformWithSecureVarFile extends TerraformCommandDecorator{
     async onRun(context: TerraformCommandContext): Promise<void> {        
         if(this.secureVarFileId){
             const secureFilePath = await this.taskAgent.downloadSecureFile(this.secureVarFileId);
-            context.terraform.arg(`-var-file=${secureFilePath}`);
+            const fileName = tasks.getSecureFileName(this.secureVarFileId);
+            if(this.isEnvFile(fileName)) {
+                let config = dotenv.config({ path: secureFilePath }).parsed;
```
Comment (charleszipp) [inline TerraformCLI/src/terraform-runner.ts]: What happens if someone uploads a .env that dotenv cannot parse?

File: README.md (hunk 1 above)
Comment (TsuyoshiUshio) [reply README.md]: Done.

File: TerraformCLI/src/terraform-validate.ts (hunk 2 above)
Comment (TsuyoshiUshio) [reply TerraformCLI/src/terraform-validate.ts]: Done

**Comment by charleszipp** [comment]
@TsuyoshiUshio I have added in the tests for the .env files. 

In order to test those, i had to update the way the secure files were tested. With the refactored test changes, `inputSecureVarFile` in the scenario definition will now set an environment variable that is used by the `azure-pipelines-task-lib/task` library to determine the name of the secure file var. This enables functions from the library like `tasks.getSecureFileName` to work when running under the context of a unit test. 
The scenario env files now need to define an actual path to an actual .env file. This is needed so that dotenv can actually load the file which in turn allows us to test the code pathways for loading an actual .env file. 
The .env files used by the tests have been setup in each relevant test directory and they are copied into the .bin directory as part of the build.

Once the build/release finishes, I should be able to merge the PR

**Pull request merged**